File: apps/mobile/lib/core/leave_dates.dart
import 'models.dart';

class CalendarDay {
  const CalendarDay({
    required this.date,
    required this.dateKey,
    required this.dayNumber,
    required this.isCurrentMonth,
  });

  final DateTime date;
  final String dateKey;
  final int dayNumber;
  final bool isCurrentMonth;
}

String dateKey(DateTime date) {
  return '${date.year.toString().padLeft(4, '0')}-'
      '${date.month.toString().padLeft(2, '0')}-'
      '${date.day.toString().padLeft(2, '0')}';
}

DateTime parseDateKey(String value) {
  final parts = value.split('-');
  if (parts.length != 3) return DateTime.now();
  final year = int.tryParse(parts[0]) ?? DateTime.now().year;
  final month = int.tryParse(parts[1]) ?? DateTime.now().month;
  final day = int.tryParse(parts[2]) ?? DateTime.now().day;
  return DateTime(year, month, day);
}

String todayKey() => dateKey(DateTime.now());

DateTime monthAfterDate(DateTime date, int deltaMonths) {
  return DateTime(date.year, date.month + deltaMonths);
}

List<CalendarDay> buildMonthGrid(DateTime monthDate) {
  final firstOfMonth = DateTime(monthDate.year, monthDate.month);
  final mondayOffset = firstOfMonth.weekday - DateTime.monday;
  final firstGridDay = firstOfMonth.subtract(Duration(days: mondayOffset));

  return List.generate(42, (index) {
    final date = firstGridDay.add(Duration(days: index));
    return CalendarDay(
      date: date,
      dateKey: dateKey(date),
      dayNumber: date.day,
      isCurrentMonth: date.month == monthDate.month,
    );
  });
}

bool isDateInRange(String value, String? startDate, String? endDate) {
  if (startDate == null || endDate == null) return false;
  return value.compareTo(startDate) >= 0 && value.compareTo(endDate) <= 0;
}

int countInclusiveDays(String startDate, String endDate) {
  final start = parseDateKey(startDate);
  final end = parseDateKey(endDate);
  return end.difference(start).inDays + 1;
}

bool selectedRangeOverlapsRequest({
  required String startDate,
  required String endDate,
  required LeaveRequest request,
}) {
  return startDate.compareTo(request.endDate) <= 0 &&
      endDate.compareTo(request.startDate) >= 0;
}

List<LeaveRequest> leaveRequestsForDate(
  String value,
  List<LeaveRequest> requests,
) {
  return requests
      .where(
        (request) =>
            request.status != LeaveRequestStatus.rejected &&
            request.startDate.compareTo(value) <= 0 &&
            request.endDate.compareTo(value) >= 0,
      )
      .toList(growable: false);
}

File: apps/web/src/lib/leaveDates.ts
import type { LeaveRequest } from "@/services/api/leaveRequestApi";

const DAY_MS = 24 * 60 * 60 * 1000;

export type CalendarDay = {
  date: Date;
  dateKey: string;
  dayNumber: number;
  isCurrentMonth: boolean;
};

export type SelectedRange = {
  startDate: string | null;
  endDate: string | null;
};

export function toDateKey(date: Date): string {
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, "0");
  const day = String(date.getDate()).padStart(2, "0");
  return `${year}-${month}-${day}`;
}

export function parseDateKey(value: string): Date {
  const [year, month, day] = value.split("-").map(Number);
  return new Date(year, month - 1, day);
}

export function todayDateKey(): string {
  return toDateKey(new Date());
}

export function addDays(date: Date, days: number): Date {
  return new Date(date.getFullYear(), date.getMonth(), date.getDate() + days);
}

export function monthAfter(date: Date, deltaMonths: number): Date {
  return new Date(date.getFullYear(), date.getMonth() + deltaMonths, 1);
}

export function buildMonthGrid(monthDate: Date): CalendarDay[] {
  const firstOfMonth = new Date(
    monthDate.getFullYear(),
    monthDate.getMonth(),
    1,
  );
  const mondayOffset = (firstOfMonth.getDay() + 6) % 7;
  const firstGridDay = addDays(firstOfMonth, -mondayOffset);

  return Array.from({ length: 42 }, (_, index) => {
    const date = addDays(firstGridDay, index);
    return {
      date,
      dateKey: toDateKey(date),
      dayNumber: date.getDate(),
      isCurrentMonth: date.getMonth() === monthDate.getMonth(),
    };
  });
}

export function isDateInRange(
  dateKey: string,
  startDate: string | null,
  endDate: string | null,
): boolean {
  if (!startDate || !endDate) return false;
  return dateKey >= startDate && dateKey <= endDate;
}

export function countInclusiveDays(startDate: string, endDate: string): number {
  const start = parseDateKey(startDate);
  const end = parseDateKey(endDate);
  return Math.floor((end.getTime() - start.getTime()) / DAY_MS) + 1;
}

export function selectedRangeOverlapsRequest(
  range: { startDate: string; endDate: string },
  request: LeaveRequest,
): boolean {
  return range.startDate <= request.endDate && range.endDate >= request.startDate;
}

export function requestsForDate(
  dateKey: string,
  requests: LeaveRequest[],
): LeaveRequest[] {
  return requests.filter(
    (request) =>
      request.status !== "REJECTED" &&
      request.startDate <= dateKey &&
      request.endDate >= dateKey,
  );
}
